### PR TITLE
Create `generate_hierarchical_df`

### LINKED ISSUE #1056: Create `generate_hierarchical_df` method
### 🚀 Feature Request

Create method to generate random hierarchical dataset

### Proposal

1. In etna/datasets/datasets_generation.py create method:
```
def generate_hierarchical_df(periods: int,  n_segments: List[int], freq: str = "D", start_time: str = "2000-01-01", ar_coef: Optional[list] = None, sigma: float = 1, random_seed: int = 1) -> pd.Dataframe
```
Parameters:
- n_segments -- number of segments on each level
- Other parameters are the same as in `generate_ar_df`

Description:
- Validate n_segments: number of segments on each level should be lower than on the next level
- Generate segments on the last level using `generate_ar_df`
- Generate random tree with  configuration of nodes on levels from n_segments
- In the dataframe replace column `segment` with columns describing the structure of the tree(one column for each level)
- Node names in the levels should be generated as follows "level_<level_id>_<segment_id>" -- you can come up with better ideas for naming 
- On the bottom level leave the default segment names from `generate_ar_df`
2. Add example of creating dataset with hierarchical structure using `generate_hierarchical_df` to the docs of TSDataset [here](https://github.com/tinkoff-ai/etna/blob/f46b363b9fffe8d0e5f52992fd655e69dcaf8683/etna/datasets/tsdataset.py#L85)

### Test cases

1. Method generate dataframe with correct properties
- number of segments
- number of periods
- columns(timestamp, level columns, target)
- level columns contains correct values(for some corner cases where randomness does not influence like n_segments=[1, 2]) 
2. Check that we can convert this dataframe to the wide format using `to_hierarchical_dataset`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1056

### CODE REVIEW

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]
You also need to create example in TSDataset, see the second task in the issue

File: tests/test_datasets/test_datasets_generation.py [hunk 1]
```diff
@@ -90,3 +92,35 @@ def test_simple_from_patterns_df_check(add_noise, checker):
...
+
+
+@pytest.mark.parametrize("periods,n_segments", ((2, [1, 2]), (2, [2]), (4, [3, 4]), (4, [3, 3])))
+def test_generate_hierarchical_df(periods, n_segments):
+    hierarchical_df = generate_hierarchical_df(periods=periods, n_segments=n_segments)
+
+    expected_columns = [f"level_{idx}" for idx in range(len(n_segments))]
+    expected_columns = set(expected_columns) | {"target", "timestamp"}
+
+    assert len(expected_columns ^ set(hierarchical_df.columns)) == 0
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply tests/test_datasets/test_datasets_generation.py:115]: Here column lists directly compared for equality of sets.

File: tests/test_datasets/test_datasets_generation.py [hunk 2]
```diff
@@ -90,3 +92,35 @@ def test_simple_from_patterns_df_check(add_noise, checker):
...
+    assert hierarchical_df["timestamp"].nunique() == periods
+
+    for level_id in range(len(n_segments)):
+        assert hierarchical_df[f"level_{level_id}"].nunique() == n_segments[level_id]
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply tests/test_datasets/test_datasets_generation.py:119]: Do you mean names of segments in each level? Is it realy necessary in case of generated names?

File: etna/datasets/datasets_generation.py [hunk 3]
```diff
@@ -186,3 +186,89 @@ def generate_from_patterns_df(
...
+        ar_coef=ar_coef,
+        sigma=sigma,
+        n_segments=n_segments[-1],
+        freq=freq,
+        random_seed=random_seed,
+    )
+
+    bottom_segments = np.unique(bottom_df["segment"])
+
+    child_to_parent = dict()
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/datasets/datasets_generation.py:247]: Current implementation is pretty straightforward. It randomly assigns each segment at the current level to parent from the next level. It looks a bit bulky because mapping builds on the fly.

Could you elaborate on point 3 of your proposal? In point 2 we are working with the number of level segments and in point 3 we want to index the entire column.

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/datasets/tsdataset.py [hunk 4]
```diff
@@ -83,6 +83,23 @@ class TSDataset:
...
+
+    >>> from etna.datasets import generate_hierarchical_df
+    >>> pd.options.display.width = 0
+    >>> n_segments = [2, 4]
+    >>> df = generate_hierarchical_df(periods=100, n_segments=n_segments, start_time="2021-01-01",)
+    >>> level_names = [f"level_{idx}" for idx in range(len(n_segments))]
+    >>> df, hierarchical_structure = TSDataset.to_hierarchical_dataset(df=df, level_columns=level_names)
+    >>> tsdataset = TSDataset(df=df, freq="D", hierarchical_structure=hierarchical_structure)
+    >>> tsdataset.df.head(5)
+    segment    level_0_0_segment_3 level_0_1_segment_0 level_0_1_segment_1 level_0_1_segment_2
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/datasets/tsdataset.py:95]: This prefixes appeared after name concatanation in `to_hierarchical_dataset`.

File: tests/test_datasets/test_datasets_generation.py [hunk 5]
```diff
@@ -90,3 +92,56 @@ def test_simple_from_patterns_df_check(add_noise, checker):
...
+def test_generate_hierarchical_df_negative_size_error(n_segments):
+    with pytest.raises(ValueError, match="All `n_segments` elements should be positive!"):
+        generate_hierarchical_df(periods=2, n_segments=n_segments)
+
+
+@pytest.mark.parametrize("periods,n_segments", ((2, [1, 2]), (2, [2]), (4, [3, 4]), (4, [3, 3])))
+def test_generate_hierarchical_df_columns_set(periods, n_segments):
+    hierarchical_df = generate_hierarchical_df(periods=periods, n_segments=n_segments)
+
+    expected_columns = [f"level_{idx}" for idx in range(len(n_segments))]
```
Comment (alex-hse-repository) [inline tests/test_datasets/test_datasets_generation.py:112]: Better to define them explicitly in the parametrization

File: tests/test_datasets/test_datasets_generation.py [hunk 6]
```diff
@@ -90,3 +92,56 @@ def test_simple_from_patterns_df_check(add_noise, checker):
...
+
+    for level_id, segment_size in enumerate(n_segments):
+        assert hierarchical_df[f"level_{level_id}"].nunique() == segment_size
+
+
+@pytest.mark.parametrize("periods,n_segments", ((2, [1, 2]), (2, [2]), (4, [3, 4]), (4, [3, 3])))
+def test_generate_hierarchical_df_segments_names(periods, n_segments):
+    hierarchical_df = generate_hierarchical_df(periods=periods, n_segments=n_segments)
+
+    num_segments = len(n_segments)
```
Comment (alex-hse-repository) [inline tests/test_datasets/test_datasets_generation.py:136]: num_levels?

File: etna/datasets/tsdataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:95]: Looks weird)))

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

**Review by alex-hse-repository — APPROVED** [review]
